src/appleseed/renderer/modeling/bsdf/bsdf.h

```diff
@@ -198,6 +198,21 @@ class APPLESEED_DLLSYMBOL BSDF
         const float                 distance,
         Spectrum&                   absorption) const;
 
+    // For layered BSDFs, compure the attenuation of substrate layers.
+    virtual void attenuate_substrate(
+        const void*                 data,                       // input values
+        const foundation::Basis3f&  shading_basis,              // world space orthonormal basis around shading normal
+        const foundation::Vector3f& outgoing,                   // world space outgoing direction, unit-length
+        const foundation::Vector3f& incoming,                   // world space incoming direction, unit-length
+        Spectrum&                   value) const;               // attenuated value
+
+    virtual void attenuate_substrate(
```

Comment:
Why do we need to expose two variants of this method?

Reply:
One is for BSDFs, the other one is used for BSSRDFs and EDFs.